gazebo/dave_ros_gz_plugins/src/DVLBridge.cc

```diff
@@ -0,0 +1,197 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iostream>
+
+#include <gz/common/Console.hh>
+#include <gz/transport/Node.hh>
+#include <rclcpp/rclcpp.hpp>
+#include "gz/plugin/Register.hh"
+
+#include "dave_ros_gz_plugins/DVLBridge.hh"
+
+GZ_ADD_PLUGIN(
+  dave_ros_gz_plugins::DVLBridge, gz::sim::System, dave_ros_gz_plugins::DVLBridge::ISystemConfigure,
+  dave_ros_gz_plugins::DVLBridge::ISystemPostUpdate)
+
+namespace dave_ros_gz_plugins
+{
+
+struct DVLBridge::PrivateData
+{
+  // Add any private data members here.
+  std::mutex mutex_;
+  gz::transport::Node gz_node;
+  std::string dvl_topic;
+  rclcpp::Publisher<dave_interfaces::msg::DVL>::SharedPtr dvl_pub;
+};
+
+DVLBridge::DVLBridge() : dataPtr(std::make_unique<PrivateData>()) {}
+
+void DVLBridge::Configure(
+  const gz::sim::Entity & _entity, const std::shared_ptr<const sdf::Element> & _sdf,
+  gz::sim::EntityComponentManager & _ecm, gz::sim::EventManager & _eventManager)
+{
+  gzdbg << "dave_ros_gz_plugins::DVLBridge::Configure on entity: " << _entity << std::endl;
+
+  if (!rclcpp::ok())
+  {
+    rclcpp::init(0, nullptr);
+  }
+
+  this->ros_node_ = std::make_shared<rclcpp::Node>("dvl_bridge_node");
+
+  // Grab dvl topic from SDF
+  if (!_sdf->HasElement("topic"))
+  {
+    this->dataPtr->dvl_topic = "/dvl/velocity";
+    gzmsg << "dvl topic set to default:  " << this->dataPtr->dvl_topic << std::endl;
+  }
+  else
+  {
+    this->dataPtr->dvl_topic = _sdf->Get<std::string>("topic");
+    gzmsg << "dvl topic: " << this->dataPtr->dvl_topic << std::endl;
+  }
+
+  // Gazebo subscriber
+  std::function<void(const gz::msgs::DVLVelocityTracking &)> callback =
+    std::bind(&DVLBridge::receiveGazeboCallback, this, std::placeholders::_1);
+
+  this->dataPtr->gz_node.Subscribe(this->dataPtr->dvl_topic, callback);
+
+  // ROS2 publisher
+  this->dataPtr->dvl_pub =
+    this->ros_node_->create_publisher<dave_interfaces::msg::DVL>(this->dataPtr->dvl_topic, 1);
+}
+
+void DVLBridge::receiveGazeboCallback(const gz::msgs::DVLVelocityTracking & msg)
+{
+  std::lock_guard<std::mutex> lock(this->dataPtr->mutex_);
+
+  gzmsg << "dave_ros_gz_plugins::DVLBridge::receiveGazeboCallback" << std::endl;
+
+  auto dvl_msg = dave_interfaces::msg::DVL();
+  dvl_msg.header.stamp.sec = msg.header().stamp().sec();
+  dvl_msg.header.stamp.nanosec = msg.header().stamp().nsec();
+
+  std::string dvl_type;
+  std::string target_type;
+  std::string reference_type;
+
+  switch (msg.type())
+  {
+    case DVLType::DVL_TYPE_UNSPECIFIED:
+      dvl_type = "DVL_TYPE_UNSPECIFIED";
+      break;
+    case DVLType::DVL_TYPE_PISTON:
+      dvl_type = "DVL_TYPE_PISTON";
+      break;
+    case DVLType::DVL_TYPE_PHASED_ARRAY:
+      dvl_type = "DVL_TYPE_PHASED_ARRAY";
+      break;
+    default:
+      dvl_type = "DVL_TYPE_UNSPECIFIED";
+  }
+
+  switch (msg.target().type())
+  {
+    case DVLTargetType::DVL_TARGET_UNSPECIFIED:
```

Comment:
Changing `case DVLTargetType::DVL_TARGET_UNSPECIFIED:` to `case gz::msgs::DVLVelocityTracking::DVL_TYPE_UNSPECIFIED:` seems to compile alright. But not sure this is right fix.

Reply:
Also, `DVL_TARGET_WATER` should be `DVL_TARGET_WATER_MASS`

Reply:
changed